### DISCUSSION TITLE
UI를 그릴때 Lazy로만 그리는게 좋다?

### DISCUSSION THREAD

**Discussion by dangsal** [DeveloperAcademy-POSTECH/MC2-Team5-Firefighter#143] [General]
## 사건의 발단
<img width="717" alt="스크린샷 2022-08-15 오후 7 58 02" src="https://user-images.githubusercontent.com/59243274/184623648-901ca356-ef18-4be4-b44b-f2f6abc7d5c4.png">

케미의 코드리뷰 중 lazy 관련 얘기가 나왔고 링고께서 UI Component를 그릴 때 lazy를 사용하는게 좋다! 라고 하셔서 의문이 들기 시작했습니다. 

그렇다면 lazy 에 대해서 부터 천천히 파헤쳐 봐봅시다!

## Lazy property

"A lazy stored property is a property whose initial value is not calculated until the first time it is used”

→ lazy stored property 는 처음 사용될 때까지 초기값이 계산되지 않는 프로퍼티 입니다.

## 왜 사용할까?

1. **메모리를 효율적으로 사용하기 위함이다.** 
일반적으로 변수를 var, let 를 사용하여 초기화한 후 선언하지만, 이러한 초기화 시점을 이 변수가 처음 참조, 이용되는 시점으로 의도적으로 지연시키는 것이다. → 어떠한 뷰가 로드 되는 시점에 모든 변수를 한번에 초기화하여 메모리에 올려두는데, 이러한 변수들은 반드시 사용되는 것은 아니므로 리소스 낭비가 될수 있다. 
2. **초기화 시점을 다르게!**
말 그대로 Lazy (지연) 프로퍼티 라는 점이다. 
iOS는 멀티쓰레드에서 초기화를 하기 때문에 의도적으로 지연시켜 줘야 할때 사용한다

## Lazy propert 사용할 때 고려사항

- lazy 는 반드시 var 와 쓰여야 한다.
기본적으로 lazy로 선언되는 변수는 초기에 값이 존재하지 않고, 이후에 값이 생기는 것이기 때문에 let 으로 선언될 수 없다.
- struct, class
기본적으로 lazy는 struct 와 class 에서만 사용할 수 있다.
- lazy vs Computed property (캐싱?)
Computed Property 에는 lazy 키워드를 사용할 수 없다. lazy는 처음 사용될 때 메모리에 값을 올리고 그 이후부터는 계속해서 메모리에 올라온 값을 사용한다. 사용할 때마다 값을 연산해서 사용하는 Computed property에는 사용할 수 없다.
- lazy and closure
lazy에는 어떤 특별한 연산을 통해 값을 넣어주기 위해 closure를 사용한다.
class 나 struct의 다른 프로퍼티으 값을 lazy 변수에서 사용하기 위해서는 closure 내에서 self를 통해 접근이 가능하다. 기본적으로 일반 변수들은 클래스가 생성된 이후에 접근이 가능하기 때문에 클래스 내의 다른 영역(메소드, 일반 프로퍼티)에서는 self를 통해 접근할 수 없지만, lazy키워드가 붙으면 생성 후 추후에 접근할 것이라는 의미이기 때문에 closure내에서 self로 접근이 가능하다.

## 그러면 Lazy를 계속 사용하는게 좋지 않나?

- lazy로만 사용한다면 뷰를 보여줄때 사용자가 느리다고 생각할 수 있다.
    - 화면을 그릴때 처음부터 사용될 UI를 lazy로 그려주는 것은 그다지 의미가 없다.
- lazy로 사용하면 thread safe 하지 못하다 → 아직 초기화되지 않은 lazy 프로퍼티에 멀티스레드가 동시에 접근하면, 프로퍼티가 한번 초기화되는것이 보장되지 않는다!
    
    > If a property marked with the `lazy`
     modifier is accessed by multiple threads simultaneously and the property has not yet been initialized, there is no guarantee that the property will be initialized only once.
    > 
    
    [[Properties - The Swift Programming Language (Swift 5.7)](https://docs.swift.org/swift-book/LanguageGuide/Properties.html)](https://docs.swift.org/swift-book/LanguageGuide/Properties.html)
    
    두 쓰레드가 동시에 액세스하려고 하면 부분적으로 초기화 된 개체를 수신할 수 있어서 안전하지 않다
    
    멀티 쓰레드 환경에서 모든 lazy 변수가 한번 초기화 되는것이 보장되지 않는것인가? → 명시하거나 static 사용해야하나?
    
    모든 Cocoa/Cocoa Touch 클래스와 메소드는 문서에서 쓰레드로부터 안전하다고 명시적으로 선언된 것을 제외하고는 스레드로 부터 안전하지 않다. (immutable 클래스는 쓰레드로 부터 안전할 것으로 기대되지만 보장되지 않음)

### 추후,...
링고께서 말씀하신 Only lazy 사용의 장점은 사실 찾을 수 가 없었습니다. 
discussion으로 올리려한 이유가 뭔가 장점을 찾고 뭐가 더 좋다 이게 더 좋다를 얘기하고 싶어서 올리려 했는데 일방적인 의견밖에 찾을 수 없어 애매해 졌네요 ㅎㅎㅎ 그래도 일단 제 생각은 이렇고 방학 기간이 끝나면 멘토분들께도 여쭤봐서 내용 추가하도록 하겠습니다!!

## Resource

- https://docs.swift.org/swift-book/LanguageGuide/Properties.html
- https://stackoverflow.com/questions/38141298/lazy-initialisation-and-retain-cycle
- https://velog.io/@minji0801/Swift-기초-문법-lazy-키워드
- https://www.avanderlee.com/swift/lazy-var-property/
- https://brunch.co.kr/@tilltue/71
- https://github.com/apple/swift/issues/43654

**Comment by skycat0212** [comment]
링고께 다시 여쭤본 내용을 급하게 정리한 것입니당

```
어떤 컴포넌트는 무조건 화면에 나오면 lazy를 쓸 필요가 없다.

오히려 미리 로드를 해서 성능을 높이는게 좋음

안쓰는게 좋음

나올수도 안나올수도 있으면…

xib나 스토리보드를 불러오면 초기화를 완전히 하지는 않지만 그 안에 있는 오브젝트들이 메모리를 어느정도 잡아먹음.

아예 얘는 화면에 보이기 전까지는 메모리 allocation도 하지마 하려면 lazy가 좋음

뷰 하나하나가 무겁던 저메모리 시절에는 lazy가 효율을 위해 필요했지만,
지금은?

이런 lazy같이 메모리 같은 애들은 환경에 따라 달라짐

뭐 ar vr 같은거면 무거우니까 lazy 한게 나음

하지만 라벨 같은거는 굳이

텍스트 필드는 엄청 클 수도 있어. 근데 그마저도 요즘에는 lazy 하지 않는 추세?

보이지 않는 오브젝트들로 인한 메모리 낭비를 줄이기 위한 것

확실히 나오는 애들은 얼리로딩이 나음

성능과 메모리!
```

**Reply by MMMIIIN** [reply]
와우 ! 스누피 정리 감사합니다 !
멋져멋져
